src/document.rs

```diff
@@ -1,27 +1,350 @@
     #[serde(skip_serializing_if = "String::is_empty")]
-    _key: String,
-
+    pub _key: String,
     #[serde(skip_serializing_if = "String::is_empty")]
-    _rev: String,
+    pub _rev: String,
+}
+
+/// Content Response when having CRUD operation on document
+#[derive(Serialize, Deserialize, Debug)]
+pub struct OperationResponse<T> {
+    #[serde(flatten)]
+    /// May contain the { _key : String, _id : String, _rev:String } of the
+    /// document
+    pub header: DocumentHeader,
+    /// May contain the document after being created/replace/updated
+    pub new: Option<T>,
+    #[serde(rename = "_oldRev")]
+    /// May contain the old revision of the document after update/replace
+    pub _old_rev: Option<String>,
+    /// May contain the old the document after update/replace/remove
+    pub old: Option<T>,
+}
+
+/// Standard Response when having CRUD operation on document
+/// Todo could add more response variant like shown on official doc
+/// 200: is returned if the document was found
+/// 304: is returned if the “If-None-Match” header is given and the document has
+/// the same version 404: is returned if the document or collection was not
+/// found 412: is returned if an “If-Match” header is given and the found
+/// document has a different version. The response will also contain the found
+/// document’s current revision in the Etag header.
+pub enum DocumentResponse<T> {
+    /// Silent is when there is empty object returned by the server
+    Silent,
+    /// Contain data after CRUD
+    Response {
+        header: DocumentHeader,
+        old: Option<T>,
+        new: Option<T>,
+        _old_rev: Option<String>,
+    },
+}
+
+// Gives extra method on the DocumentResponse to quickly check what the server
+/// returns
+impl<T> DocumentResponse<T> {
+    /// Should be true when the server send back an empty object {}
+    pub fn is_silent(&self) -> bool {
+        match self {
+            DocumentResponse::Silent => true,
+            _ => false,
+        }
+    }
+    /// Should be true if there is a response from the server
+    pub fn has_response(&self) -> bool {
+        match self {
+            DocumentResponse::Response { .. } => true,
+            _ => false,
+        }
+    }
+
+    /// Should give None or Some(Response)
+    pub fn get_response(self) -> Option<OperationResponse<T>> {
+        if let DocumentResponse::Response {
+            header,
+            old,
+            new,
+            _old_rev,
+        } = self
+        {
+            Some(OperationResponse {
+                header,
+                old,
+                new,
+                _old_rev,
+            })
+        } else {
+            None
+        }
+    }
+}
+
+impl<'de, T> Deserialize<'de> for DocumentResponse<T>
+where
+    T: Deserialize<'de>,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let map = serde_json::Map::deserialize(deserializer)?;
```

Comment:
The implementation can be improved a little bit:

1. deserialize it to `Value` in the first place instead of `Map`, and use [`as_object()`](https://docs.serde.rs/serde_json/enum.Value.html#method.as_object) to see if it contains some keys.
2. use [`remove](https://docs.serde.rs/serde_json/map/struct.Map.html#method.remove)` to get the value of `old`, `_old_rev` and `get`, which can avoid clone when deserialize into `DocumentHeader`,